File: assets/ws/generate-ts-client/sample/lib/base/IWSBase.ts
import { SimpleEventDispatcher, SignalDispatcher } from 'strongly-typed-events';
import { IEventError } from './IEventError';
import { IEventData } from './IEventData';
import { INestJSWSException } from './INestJSWSException';

export interface IWSBase {
    //#region  [ properies ]
    readonly id: string | undefined;
    defaultRequestTimeout: number;
    readonly isConnected: boolean;
    readonly baseUrl: string | undefined;
    readonly path: string | undefined;
    readonly nsp: string | undefined;
    //#endregion

    //#region  [ events ]
    onConnectionChange: SimpleEventDispatcher<boolean>;
    onSend: SimpleEventDispatcher<IEventData>;
    onReceive: SimpleEventDispatcher<IEventData>;
    onNewSocketInstance: SignalDispatcher;
    onSubscriptionError: SimpleEventDispatcher<IEventError>;
    onNestJSException: SimpleEventDispatcher<INestJSWSException>;
    //#endregion

    //#region  [ methods ]
    connect(baseUrl: string, path: string, nsp: string, authToken?: string): void;
    disconnect(): void;

    subscribe(event: string, action: (data: any) => void): IWSBase;
    publish(event: string, data?: any): void;

    requestAsync<T = void>(event: string, data?: any, timeout?: number): Promise<T>;
    //#endregion
}

File: assets/ws/generate-cs-client/sample/src/template.api.wsclient/src/lib/ws/base/IWSBase.cs
﻿using Newtonsoft.Json.Linq;
using nex.types;
using System;
using System.Threading.Tasks;

namespace nex.ws
{
    public interface IWSBase
    {
        #region [ properties ]
        string Id { get; }
        int DefaultRequestTimeout { get; set; }
        bool IsConnected { get; }
        string Url { get; }
        string Nsp { get; }
        #endregion

        #region [ events ]
        event EventHandler<EventArgs<EventData>> EventSend;
        event EventHandler<EventArgs<EventData>> EventReceive;
        event EventHandler<EventArgs<bool>> EventConnectionChange;
        event EventHandler<EventArgs> EventNewSocketInstance;
        event EventHandler<EventArgs<EventError>> EventSubscriptionError;
        event EventHandler<EventArgs<NestJSWSException>> EventNestJSException;
        #endregion

        #region [ methods ]
        void Connect(string url, string nsp, string authToken = null);
        void Disconnect();

        IWSBase Subscribe(string eventName, Action<JToken> action);
        void Publish(string eventName, object data = null);

        Task RequestAsync(string eventName, object data, int timeout);
        Task<T> RequestAsync<T>(string eventName, object data, int timeout);
        #endregion
    }
}